src/dsl/parser.rs

```diff
@@ -57,29 +59,95 @@ impl<'input> Parser<'input> {
         Ok(rhs)
     }
 
-    fn parse_rhs_impl(&mut self) -> Result<Rhs> {
-        let mut entries = Vec::new();
+    fn parse_rhs_impl(&mut self, depth: usize, return_on_dot: bool) -> Result<Rhs> {
+        if depth > MAX_DEPTH {
+            return Err(ParseError {
+                pos: self.input.pos(),
+                cause: ParseErrorCause::MaximumRecursion(MAX_DEPTH).into(),
+            });
+        }
+
+        let mut parts = Vec::new();
+
+        let mut last_was_dot = false;
 
         while let Some(token) = self.input.peek() {
             let token = token?;
-            let res = match &token.kind {
-                TokenKind::Amp => self.parse_amp().map(|t| RhsEntry::Amp(t.0, t.1)),
-                TokenKind::At => self.parse_at().map(RhsEntry::At),
-                TokenKind::OpenBrkt => self.parse_index_op().map(RhsEntry::Index),
+            let part = match &token.kind {
+                TokenKind::OpenBrkt => {
+                    let idx_op = self.parse_index_op(depth)?;
+                    RhsPart::Index(idx_op)
+                }
+                TokenKind::Amp | TokenKind::At | TokenKind::Key(_) => self.parse_rhs_part(depth)?,
                 TokenKind::Dot => {
+                    if return_on_dot {
+                        return Ok(Rhs(parts));
+                    }
+
+                    // can't start with a dot and can't have consecutive dots
+                    if parts.is_empty() || last_was_dot {
+                        return Err(ParseError {
+                            pos: token.pos,
+                            cause: ParseErrorCause::UnexpectedToken(
+                                self.input.next().unwrap().unwrap(),
```

Comment:
this would never panic because the code calls `peek` before

Reply:
Calling unwrap is not necessarily a bad thing. Even tokio has a bunch of unwraps in the code when the code does some irrecoverable thing. In this case this thing should never panic so it is better/ok to express in the code directly in my opinion. On top of this we have a ton of tests and fuzz testing as well which catches these panics if we were to introduce that kind of panic.

For example if we were to introduce that bug, which errored here. It would be harder to debug it if it was returning a unexpected end of input error or some weird thing like that here.

Reply:
The well-designed library has no right to panic (unless it is very explicit in API, e.g. `_unchecked` suffix or another indication ) because it might be used by an application that operates very sensitive data, and the devs of that application will not be happy that our library crashed the process. 

> it is better/ok to express in the code directly in my opinion
How is it expressed exactly? A reader has to be deeply in the context to say that it would never panic.

Reply:
Replacing all of this with error handling

Reply:
Actually I think it should be ok still because it is a common pattern in code and it is very easily catched by fuzz testing if we were to introduce a panic like this.

https://users.rust-lang.org/t/unwrap-in-libraries/15558

Reply:
This link doesn't contradict the point I am trying to make.
Anyway,  I leave it to @sehz  to decide. I, personally, have a rule of thumb - never use unwraps or panic in the code other than the `main` function in rear cases. But that should definitely be up to the code convention accepted by the team.

Reply:
I tended to agree with @galibey.   However, there are a few exceptions, like mutex poisoning, assertions, OOM, and unsafe code exception, where it is costly to build exception handling, given it will not occur in practice.   Generally, whenever I see code that uses raw unwraps (without defaults), it lacks abstraction to handle edge cases.

So @ozgrakkurt, should explain why `unwrap` is warranted and whether there might be different ways to express abstraction so as not to cause `unwrap.`